src/parser.h

```diff
@@ -0,0 +1,32 @@
+#ifndef ADA_PARSER_H
+#define ADA_PARSER_H
+
+#include <optional>
+
+namespace ada {
+
+  class Parser {
+
+  public:
+    Parser(const char* input, std::optional<ada::URL> optional_base_url, std::optional<ada::encoding_type> encoding, std::optional<ada::state> state);
```

Comment:
You take in a `const char*` input and then cast away the `const` qualifier.  Some people will object to such an approach.

Reply:
I converted this to `string_view`. I believe even though `const std::string_view`, we do not need to drop the const  qualifier, and now can just iterate through the `const std::iterator`.